ahl_health_apps_analysis/pipeline/cluster.py

```diff
@@ -0,0 +1,149 @@
+		logging.info(
+			f"{len(set(clusterer.labels_))} clusters with {sum([bool(x) for x in clusterer.labels_ if int(x) == -1])} unassigned apps"
+		)
+
+		# step 4 - 
+		df = details.copy()
+		df["x"] = embedding[:, 0]
+		df["y"] = embedding[:, 1]
+		df["cluster"] = [int(x) for x in clusterer.labels_]
+		df["cluster_confidence"] = [str(x) for x in clusterer.probabilities_]
+
+
+		# step 5 - 
+		cluster_col = df.columns.get_loc("cluster")
+
+		radius = 3
+		neigh = KNeighborsClassifier(n_neighbors=radius)
+
+		current_unassigned = len(df[df.cluster == -1].cluster.to_list())
+
+		cluster_list = df.cluster.to_list()
+
+		unassigned_indices = []
+		unassigned_indices.extend([i for i, x in enumerate(cluster_list) if x == -1])
+		logging.info(f"{len(unassigned_indices)} apps not in clusters")
+
+		X_list = [x for i, x in enumerate(embedding_clustering) if i not in unassigned_indices]
+		y_list = [i for i in cluster_list if i != -1]
+
+		neigh.fit(X_list, y_list)
+
+		for x in unassigned_indices:
+			df.iat[x, cluster_col] = neigh.predict([embedding_clustering[x]])
+
+		print(f"{len(df[df.cluster == -1].cluster.to_list())} unassigned apps remain.")
+
+		#step 6
+		cluster_description = {}
+		for i, cluster in df.groupby('cluster'):
+		    cluster_description[i] = ' '.join(cluster['description'])
+
+		X = tfidf.fit_transform(cluster_description.values())
+		feature_names = np.array(tfidf.get_feature_names())
+
+		cluster_names = {i:"-".join(get_top_tf_idf_words(response,2,["app", "apps","blood"])) for i, response in enumerate(X)}
+		df['cluster_names'] = df['cluster'].map(cluster_names)
+
+		# %%
+		# Visualise using altair (NB: -1=points haven't been assigned to a cluster)
+		fig = (
+			alt.Chart(df.reset_index(), width=725, height=725)
+			.mark_circle(size=60)
+			.encode(x="x", y="y", tooltip=["cluster_names", "appId", "summary"], color="cluster_names:N")
+		).interactive()
+
+		fig.save(f'outputs/figures/{date}-hdbscan-cluster.html')
+
+	
+	elif args.cluster_type == "kmeans":
+
+		clusterer = KMeans(n_clusters=15)
+		y_predicted = clusterer.fit_predict(embedding).tolist()
+
+		# step 4 - 
+		df = details.copy()
+		df["x"] = embedding[:, 0]
+		df["y"] = embedding[:, 1]
+		df["cluster"] = [int(x) for x in y_predicted]
+
+
+		cluster_description = {}
+		for i, cluster in df.groupby('cluster'):
+		    cluster_description[i] = ' '.join(cluster['description'])
+
+		X = tfidf.fit_transform(cluster_description.values())
+		feature_names = np.array(tfidf.get_feature_names())
+
+		cluster_names = {i:"-".join(get_top_tf_idf_words(response,2,["app", "apps","blood"])) for i, response in enumerate(X)}
+		df['cluster_names'] = df['cluster'].map(cluster_names)
+		
+		fig = (
+    	alt.Chart(df.reset_index(), width=725, height=725)
+    	.mark_circle(size=60)
+    	.encode(x="x", y="y", tooltip=["cluster_names", "appId", "summary"], color="cluster_names:N")
+		).interactive()
+
+		fig.save(f'outputs/figures/{date}-kmeans-cluster.html')
```

Comment:
it looks like all these lines of code are the same in hbscan and kmeans? if so, have the cluster naming and plot building after the `if else` - that way it only needs to be written out once and means you don't need to edit code in two places if you make changes. I guess you will need another `if/else` right at the end with 
```
if :
  fig.save(f'outputs/figures/{date}-hdbscan-cluster.html')
else:
  fig.save(f'outputs/figures/{date}-kmeans-cluster.html')
 ```